File: Graphs/Graphs_bfs.cpp
// Graphs Adjacency List implementation for Generic Data
// Program Author : Abhisek Kumar Gupta
#include<bits/stdc++.h>
using namespace std;

template<typename T>
class Graph{
    map<T, list<T> > adjList;
    public:
        Graph(){

        }
        void addEdge(T u, T v, bool bidir = true){
            adjList[u].push_back(v);
            if(bidir){
                adjList[v].push_back(u);
            }
        }
        void printAdjList(){
            for(auto obj : adjList){
                cout << obj.first <<  "->";
                for(auto element : obj.second){
                    cout << element << ",";
                }
                cout << endl;
            }
        }
        void bfs(int n){
            queue<int> q;
            q.push(n);
            map<int, bool> visited;
            visited[n] = true;
            while(!q.empty()){
                int node_element = q.front();
                cout << node_element << " ";
                q.pop();
                for(int neighbour : adjList[node_element]){
                    if(!visited[neighbour]){
                        q.push(neighbour);
                        visited[neighbour] = true;
                    }
                }
            }

        }
};
int main(){
    Graph<int> g;
    g.addEdge(0, 1);
    g.addEdge(0, 4);
    g.addEdge(1, 4);
    g.addEdge(1, 3);
    g.addEdge(1, 2);
    g.addEdge(2, 3);
    g.addEdge(4, 3);
    g.printAdjList();
    g.bfs(0);
    return 0;
}

File: Graphs/Graphs_bfs.py
from collections import defaultdict

class Graph:
    def __init__(self):
        self.graph = defaultdict(list)

    def insertEdge(self,v1,v2):
        self.graph[v1].append(v2)

    def subBfs(self,visited,queue):
        while(queue):
            v=queue.pop(0)
            visited.add(v)
            print(v,end=" ")
            
            for neighbour in self.graph[v]:
                if neighbour not in visited:
                    queue.append(neighbour)
                    visited.add(neighbour)
        
    def bfs(self,v):
        visited=set()
        queue=[v]
        self.subBfs(visited,queue)

g=Graph()
g.insertEdge(1,2)
g.insertEdge(2,1)
g.insertEdge(2,3)
g.insertEdge(3,4)
g.insertEdge(4,5)
g.insertEdge(5,6)
g.insertEdge(2,6)

g.bfs(1)
